### PR TITLE
add usb device sub case - start vm with usb controller alias configured

### PR DESCRIPTION
**Pull request by brucepatricklee**
this case is based on bug
	https://bugzilla.redhat.com/show_bug.cgi?id=1552127

configure usb controllers with alias and check the name of controller,
masterbus and device id

Signed-off-by: Jin Li <jil@redhat.com>

### CODE REVIEW

**Comment by brucepatricklee** [comment]
(p3-ve) [root@rhel8 code]#  avocado run --vt-type libvirt usb_device.passthrough.bus_dev.usb_all --vt-machine-type q35
JOB ID     : 8b245767701f2b8edf2da2ecb713df1cc755175c
JOB LOG    : /root/avocado/job-results/job-2019-05-07T16.37-8b24576/job.log
 (1/1) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb_all.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (67.65 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 69.28 s

File: libvirt/tests/src/usb_device.py [hunk 1]
```diff
@@ -168,12 +189,21 @@ def usb_device_check(session, src_host=None):
...
-        for model in usb_model.split(','):
+        for i, model in enumerate(usb_model.split(',')):
             controller = Controller("controller")
             controller.type = "usb"
             controller.index = usb_index
             controller.model = model
+            if usb_alias:
+                alias = "ua-usb" + str(i) + random_id
+                controller_alias[model] = alias
+                alias = {"name": alias}
```
Comment (chloerh) [inline libvirt/tests/src/usb_device.py]: I don't thinks it's best to change type of a var like that, maybe we could try another name for that var

File: libvirt/tests/src/usb_device.py [hunk 2]
```diff
@@ -149,6 +150,26 @@ def usb_device_check(session, src_host=None):
...
 
+    def check_controller_alias(alias):
+        """
+        :param alias: a {model:alias} dict of the usb controller
+        """
+        output = process.run("ps -ef | grep {}".format(vm_name), shell=True).stdout_text
+        logging.debug('"ps -ef | grep {}" output {}'.format(vm_name, output))
+        for model in usb_model.split(','):
+            if model != "qemu-xhci":
+                device = model.split('-')[0] + '-' + "usb" + '-' + model.split('-')[1]
```
[the remark this answers is not in the document]
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: yes, that's a more beautiful one, updated

File: libvirt/tests/src/usb_device.py [hunk 3]
```diff
@@ -149,6 +150,26 @@ def usb_device_check(session, src_host=None):
...
+            else:
+                device = model
+            pattern = "-device {},".format(device)
+            if "ich9-uhci" in model:
+                pattern += "masterbus={}".format(alias['ich9-ehci1'])
+            else:
+                pattern += "id={}".format(alias[model])
```
[the remark this answers is not in the document]
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: good point, updated

File: libvirt/tests/src/usb_device.py [hunk 4]
```diff
@@ -149,6 +150,26 @@ def usb_device_check(session, src_host=None):
...
+        """
+        :param alias: a {model:alias} dict of the usb controller
+        """
+        output = process.run("ps -ef | grep {}".format(vm_name), shell=True).stdout_text
+        logging.debug('"ps -ef | grep {}" output {}'.format(vm_name, output))
+        for model in usb_model.split(','):
+            if model != "qemu-xhci":
+                device = model.split('-')[0] + '-' + "usb" + '-' + model.split('-')[1]
+            else:
+                device = model
```
[the remark this answers is not in the document]
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: good point, updated

File: libvirt/tests/src/usb_device.py (hunk 1 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: good point, updated

**Comment by qiankehan** [comment]
An error occurs when running the case:
 (1/8) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pci-root: ERROR (43.29 s)

2019-05-15 10:59:49,322 stacktrace       L0041 ERROR|
2019-05-15 10:59:49,323 stacktrace       L0044 ERROR| Reproduced traceback from: /root/avocado/avocado/core/test.py:864
2019-05-15 10:59:49,351 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-05-15 10:59:49,352 stacktrace       L0047 ERROR|   File "/root/avocado-vt/avocado_vt/test.py", line 297, in runTest
2019-05-15 10:59:49,352 stacktrace       L0047 ERROR|     raise self.__status  # pylint: disable=E0702
2019-05-15 10:59:49,352 stacktrace       L0047 ERROR| AttributeError: Key 'type' not found in super class attributes or in ('hostdev_type', 'mode', 'managed', 'sgio', 'rawio', 'source', 'boot_order', 'readonly', 'shareable', 'type_name', 'device_tag', 'xml', 'virsh', 'xmltreefile', 'validates', 'get_type', 'set_type', 'del_type')
2019-05-15 10:59:49,352 stacktrace       L0048 ERROR|
2019-05-15 10:59:49,352 test             L0869 DEBUG| Local variables:
2019-05-15 10:59:49,431 test             L0872 DEBUG|  -> self <class 'avocado_vt.test.VirtTest'>: 1-type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pci-root
2019-05-15 10:59:49,431 test             L0289 DEBUG| DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
2019-05-15 10:59:49,431 test             L0289 DEBUG| DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
2019-05-15 10:59:49,431 test             L0289 DEBUG| DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
2019-05-15 10:59:49,431 test             L1019 ERROR| Traceback (most recent call last):

2019-05-15 10:59:49,431 test             L1019 ERROR|   File "/root/avocado/avocado/core/test.py", line 955, in _run_avocado
    raise test_exception

@brucepatricklee Please check this issue

**Comment by brucepatricklee** [comment]
@qiankehan please update the avocado-vt repo, the name attribute of hostdev was changed

**Comment by brucepatricklee** [comment]
(p3-ve) [root@rhel8 code]# avocado run --vt-type libvirt usb_device --vt-machine-type q35
JOB ID     : 51e28c834341b51b3e913d08f0461e8a18140c40
JOB LOG    : /root/avocado/job-results/job-2019-05-16T17.35-51e28c8/job.log
 (01/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (67.97 s)
 (02/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (72.08 s)
 (03/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.dmi-to-pci-bridge: PASS (71.16 s)
 (04/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (71.57 s)
 (05/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (75.38 s)
 (06/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.dmi-to-pci-bridge: PASS (70.52 s)
 (07/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (74.95 s)
 (08/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (76.72 s)
 (09/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.dmi-to-pci-bridge: PASS (71.90 s)
 (10/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (85.59 s)
 (11/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (74.77 s)
 (12/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.dmi-to-pci-bridge: PASS (76.08 s)
 (13/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (75.09 s)
 (14/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (70.12 s)
 (15/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.dmi-to-pci-bridge: PASS (70.02 s)
 (16/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (84.33 s)
 (17/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (78.75 s)
 (18/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.dmi-to-pci-bridge: PASS (79.20 s)
 (19/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb_all.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (83.91 s)
 (20/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.spicevmc.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (23.62 s)
 (21/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.tcp.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (49.76 s)
RESULTS    : PASS 21 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1507.24 s

**Comment by qiankehan** [comment]
@brucepatricklee  
Works for me:
avocado run --vt-type libvirt usb_device      
JOB ID     : 8236b182fd7a39e51f285201efafbcfd9fa5869c
JOB LOG    : /root/avocado/job-results/job-2019-05-17T10.38-8236b18/job.log
 (1/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pci-root: PASS (76.42 s)
 (2/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pci-root: PASS (77.79 s)
 (3/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pci-root: PASS (82.44 s)
 (4/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pci-root: PASS (88.49 s)
 (5/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pci-root: PASS (79.43 s)
 (6/6) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pci-root: PASS (84.83 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 494.45 s


Remember to add the avocado-vt requirement to commit msg

File: libvirt/tests/src/usb_device.py [hunk 5]
```diff
@@ -149,6 +150,22 @@ def usb_device_check(session, src_host=None):
         if passthrough:
             usb_disk_check(session, src_guest)
 
+    def check_controller_alias(alias):
+        """
+        :param alias: a {model:alias} dict of the usb controller
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: Add description for the function

Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: updated

**Pull request merged**